### PR TITLE
fix error in parallel computation

### PR DESCRIPTION
**Pull request by komorihi**
To fix the issues #58, #60, and #65, I modified the code of objective.py by using joblib.

There was also a way to use `multiprocessing.Pool`, but the overhead at optimization seems to be large, which took about 10 to 20% more time than joblib.

I'm not familiar with parallel processing, so there may be better implementation methods. If anyone has a good idea, I'd be happy if you give me advice.

This is the alternative code I wrote with `multiprocessing.Pool`. (I did not include in this fix)
```python
    def _syncronous_batch_evaluation(self,x):
        from multiprocessing import Pool
        divided_samples = [x[i::self.n_procs] for i in range(self.n_procs)]
        p = Pool()
        results = p.map(self._eval_func, divided_samples)
        p.close()
        p.join()

        f_evals = np.vstack([res[0] for res in results])
        cost_evals = np.hstack([res[1] for res in results])

        return f_evals, cost_evals
```

Also note that those who do parallel calculations need to place the execution part of the main script within the `if __name__ == '__main__':` statement. If you do not do this, the main script will run infinitely.

### CODE REVIEW

**Comment by javiergonzalezh** [comment]
Hi @komorihi,

This is great. I see that some of the tests are failing for Python 2.7. Can you have a look to that before we can merge?

Thanks for contributing!

**Comment by komorihi** [comment]
The failure is caused by python2.7 can't pickle the evaluate function that is an instance method.

Even if the evaluation function is converted into a non-instance method, if the objective function wrapped by the evaluation function is an instance method, it can't pickle.

The only work around that I could find is one using multiprocess module, which is a fork of multiprocessing.

**Comment by javiergonzalezh** [comment]
@apaleyes: I will merge this one unless you have any comment. All the test pass now.

File: GPyOpt/core/task/objective.py [hunk 1]
```diff
@@ -82,22 +71,23 @@ def _syncronous_batch_evaluation(self,x):
...
+            from multiprocess import Pool
+            p = Pool(processes=self.n_procs)
+            results = p.map(self._eval_func, divided_samples)
+            p.close()
+            p.join()
+        else:
+            from joblib import Parallel, delayed
+            results = Parallel(n_jobs=self.n_procs)([delayed(self._eval_func)(args) for args in divided_samples])
+
+        f_evals = np.vstack([res[0] for res in results])
```
Comment (apaleyes) [inline GPyOpt/core/task/objective.py]: Can't this be just `np.array()`?

File: GPyOpt/core/task/objective.py [hunk 2]
```diff
@@ -46,17 +42,10 @@ def evaluate(self, x):
...
         else:
-            try:
-                f_evals, cost_evals = self._syncronous_batch_evaluation(x)
-            except:
-                if not hasattr(self, 'parallel_error'):
-                    print('Error in parallel computation. Fall back to single process!')
-                else:
-                    self.parallel_error = True
-                f_evals, cost_evals = self._eval_func(x)
+            f_evals, cost_evals = self._syncronous_batch_evaluation(x)
```
Comment (apaleyes) [inline GPyOpt/core/task/objective.py]: While we are here, would you mind fixing the typo in sync_h_ronous? Thanks!

File: GPyOpt/core/task/objective.py [hunk 3]
```diff
@@ -82,22 +71,23 @@ def _syncronous_batch_evaluation(self,x):
...
+            p = Pool(processes=self.n_procs)
+            results = p.map(self._eval_func, divided_samples)
+            p.close()
+            p.join()
+        else:
+            from joblib import Parallel, delayed
+            results = Parallel(n_jobs=self.n_procs)([delayed(self._eval_func)(args) for args in divided_samples])
+
+        f_evals = np.vstack([res[0] for res in results])
+        cost_evals = np.hstack([res[1] for res in results])
```
Comment (apaleyes) [inline GPyOpt/core/task/objective.py]: Same somment - np.array and transpose should do the job, if I am not mistaken. hstack and vstack are now considered deprecated in NumPy and are only supported for compatibility reasons.

File: requirements.txt [hunk 4]
```diff
@@ -12,6 +12,8 @@ six>=1.10.0
 mock>=2.0.0
 PyDOE >= 0.3.0
 sobol_seq >=0.1
+joblib>=0.11
```
Comment (apaleyes) [inline requirements.txt]: Can you also update setup.py with these?

File: travis_tests.py [hunk 5]
```diff
@@ -5,4 +5,6 @@
 import nose, warnings
 with warnings.catch_warnings():
     warnings.simplefilter("ignore")
-nose.main('GPyOpt', defaultTest='GPyOpt/testing/', argv=[''])
\ No newline at end of file
+
+if __name__=='__main__': # for parallel computation
```
Comment (apaleyes) [inline travis_tests.py]: I don't understand why this change was needed. Can you please elaborate here?

Comment (samischyvinck) [reply travis_tests.py]: This is the reason why multiprocessing is currently not working under windows. On Windows many types from multiprocessing need to be picklable so that child processes can use them. However, one should generally avoid sending shared objects to other processes using pipes or queues. Instead you should arrange the program so that a process which needs access to a shared resource created elsewhere can inherit it from an ancestor process

**Comment by ekalosak** [comment]
@apaleyes @javiergonzalezh This is an old PR - what's the status? Does it need to be brushed up or is it too far out of date? Is Python 2.7 still supported?

**Comment by apaleyes** [comment]
Closing as we have declared end of maintenance.
